src/main/java/com/epam/ta/reportportal/ws/rabbit/LaunchReporterConsumer.java

```diff
@@ -21,48 +21,104 @@
 import com.epam.ta.reportportal.core.launch.FinishLaunchHandler;
 import com.epam.ta.reportportal.core.launch.StartLaunchHandler;
 import com.epam.ta.reportportal.util.ProjectExtractor;
+import com.epam.ta.reportportal.ws.model.BulkRQ;
 import com.epam.ta.reportportal.ws.model.FinishExecutionRQ;
 import com.epam.ta.reportportal.ws.model.launch.StartLaunchRQ;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.amqp.rabbit.annotation.RabbitListener;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.messaging.handler.annotation.Header;
 import org.springframework.messaging.handler.annotation.Payload;
 import org.springframework.stereotype.Component;
 import org.springframework.transaction.annotation.Transactional;
 
+import java.util.List;
+import java.util.Map;
+
+import static com.epam.ta.reportportal.core.configs.RabbitMqConfiguration.DEAD_LETTER_MAX_RETRY;
+
 /**
  * @author Pavel Bortnik
  */
 @Component
 @Transactional
 public class LaunchReporterConsumer {
 
+	private static final Logger LOGGER = LoggerFactory.getLogger(LaunchReporterConsumer.class);
+
 	private DatabaseUserDetailsService userDetailsService;
 
 	private StartLaunchHandler startLaunchHandler;
 
 	private FinishLaunchHandler finishLaunchHandler;
 
 	@Autowired
-	public LaunchReporterConsumer(DatabaseUserDetailsService userDetailsService, StartLaunchHandler startLaunchHandler,
-			FinishLaunchHandler finishLaunchHandler) {
+	public LaunchReporterConsumer(DatabaseUserDetailsService userDetailsService,
+								  StartLaunchHandler startLaunchHandler,
+								  FinishLaunchHandler finishLaunchHandler) {
 		this.userDetailsService = userDetailsService;
 		this.startLaunchHandler = startLaunchHandler;
 		this.finishLaunchHandler = finishLaunchHandler;
 	}
 
-	@RabbitListener(queues = "#{ @startLaunchQueue.name }")
+	@RabbitListener(queues = "#{ @launchStartQueue.name }")
 	public void onStartLaunch(@Payload StartLaunchRQ rq, @Header(MessageHeaders.USERNAME) String username,
 			@Header(MessageHeaders.PROJECT_NAME) String projectName) {
 		ReportPortalUser userDetails = (ReportPortalUser) userDetailsService.loadUserByUsername(username);
 		startLaunchHandler.startLaunch(userDetails, ProjectExtractor.extractProjectDetails(userDetails, projectName), rq);
 	}
 
-	@RabbitListener(queues = "#{ @finishLaunchQueue.name }")
+	@RabbitListener(queues = "#{ @launchFinishQueue.name }")
 	public void onFinishLaunch(@Payload FinishExecutionRQ rq, @Header(MessageHeaders.USERNAME) String username,
-			@Header(MessageHeaders.PROJECT_NAME) String projectName, @Header(MessageHeaders.LAUNCH_ID) Long launchId) {
+			@Header(MessageHeaders.PROJECT_NAME) String projectName, @Header(MessageHeaders.LAUNCH_ID) String launchId,
+			@Header(required = false, name = MessageHeaders.XD_HEADER) List<Map<String, ?>> xdHeader) {
+		if (xdHeader != null) {
+			long count = (Long) xdHeader.get(0).get("count");
```

Comment:
could it be NPE here?

Reply:
Shouldn't be. xHeader will be null with very first normal run.
When message goes to dead-letter queue on exception, x-dead header is appended, and it keeps array of elements for each iteration. Checked it locally